### ISSUE TITLE
Inputs configuration

### ISSUE THREAD

**Issue by Kevin-Ford** [turgu1/indigo-robert-8channel-relay.indigoPlugin#1]
How do you add inputs? There is an 'Add Input' button on the configuration window but I do not see anything added when I click it

**Comment by turgu1** [comment]
When you press Add Relay or Add Input, you will see new entries appearing in the Indigo list of entries. The IP address is followed with "r<number>" for relais and "i<number>" for Inputs.

**Comment by Kevin-Ford** [comment]
Note: I updated indigo to 2021.1 on Friday

I am not sure of the keystrokes, but I managed to get one input (mapped to input 1) added to the Device.(after many opening and closing of the device, restarting the plug in, etc)  I cannot seem to add anymore...

The configuration window states 'Device Group (8 Max)'. Does this include the inputs ?  Currently I have all 8 relays mapped and 1 Input mapped and the all function as designed.

**Comment by turgu1** [comment]
You can have 8 Inputs and 8 Relays mapped. When you add an Input, you must then go to it's configuration panel to associate it to one of the 8 Inputs shown in the list.  

Can you take a snapshot of the Indigo' s list of device entries, (a part of the list that show the Input that you added  would be sufficient) and attach it to a new comment here. Thanks.

**Comment by Kevin-Ford** [comment]
<img width="1535" alt="2021-06-20_08-11-52" src="https://user-images.githubusercontent.com/51087349/122679352-57531500-d19f-11eb-88e0-595f314462ae.png">

Here you go

**Comment by turgu1** [comment]
Thanks! I will be back soon. I have to ask my friend to test it for me (I don't have Indigo myself). We were able to add input devices this morning without any problem in his installation. Will check for the Indigo version...

**Comment by turgu1** [comment]
Which version are you using? My friend is using 7.5.0

**Comment by Kevin-Ford** [comment]
I am using Indigo version 2021.1 (just released this week)

<img width="500" alt="2021-06-20_09-06-22" src="https://user-images.githubusercontent.com/51087349/122680969-d9930780-d1a6-11eb-8ddc-a6c996bbc8a9.png">

**Comment by Kevin-Ford** [comment]
I see the issue...  There can only be 9 rows in the Device Group grid.  I can add any combination of inputs and relays but after 9 rows it stops accepting any additional Devices.
<img width="420" alt="2021-06-20_09-23-32" src="https://user-images.githubusercontent.com/51087349/122681487-6048e400-d1a9-11eb-89ad-f2b95a048dbe.png">

<img width="1098" alt="2021-06-20_09-25-34" src="https://user-images.githubusercontent.com/51087349/122681516-7d7db280-d1a9-11eb-84c9-c9f1df0aa066.png">

Here is an example of 5 relays and 4 inputs...

**Comment by turgu1** [comment]
My friend did updated his version of Indigo to 2021.1. He was then able to add Input devices without any issue. There was already 8 relays defined.
![image](https://user-images.githubusercontent.com/174329/122683405-df9bdf00-d1cc-11eb-8327-54ed98b530ef.png)

**Comment by turgu1** [comment]
Is there anything that appears in the log when you tried to add a 10th device?

**Comment by Kevin-Ford** [comment]
Nothing appears in the log...  

I have the original 8 channel plugin controlling an original network relay and for fun I tried to add more inputs to it and it has the same behaviour....  Which indicates to me it may be something in my indigo....

What number is shown on the other indigo under the Configuration window? Both versions of my plugins say (8 Max).

<img width="489" alt="2021-06-20_10-51-47" src="https://user-images.githubusercontent.com/51087349/122683743-19152000-d1b6-11eb-896d-e0b4cf3ad18f.png">


 Could the plugin be revised so all 8 relays and 8 inputs are available without configuration as many Multifunction devices are?

**Comment by turgu1** [comment]
This is part of the label. It shows the same thing here. This is to indicate that you can't add more than 8 relays or 8 inputs, as there is not more than 8 inputs/8 outputs physically on the relays board.

**Comment by turgu1** [comment]
But you were able to have 9 devices defined isn't it?

**Comment by Kevin-Ford** [comment]
Yes nine devices. Almost like the plug-in has a devices definition set at 8 (with the plugin’s device IDs going from 0 - 8 giving 9 useable devices)

Does the other system have the original plug in running?

**Comment by Kevin-Ford** [comment]
What is the Device Group (XX Max) number on your friend's indigo?

**Comment by turgu1** [comment]
To my knowledge, he is running the same version of the plugin. He is not available at the moment to answer your last question.

**Comment by turgu1** [comment]
He got the same value:: 8 Max. As indicated, this is a hard coded label in the device's Python code.

I suspect that the issue is related to the sequence of action taken by you and my friend that are different. I asked him to detail the sequence he did to generate a new input device. Can you try to describe the sequence you are using to try to create a new input device? We can then compare.

**Comment by Kevin-Ford** [comment]
No matter which way I create the devices I can only have 9 created for a Relay Board. 

Today I created a new install of indigo on a different server and created a new database.  I only added the Simple 8 Channel Relay plugin to the database and I am observing the same behaviour.

My creations steps are :
1. in the database window --> DEVICES grid --> Click the 'New' button
2. in the New Device window, select 'Simple 8 Channel Relay' from the Type: dropdown
3. in the Configure Simple 8 Channel Relay Board window I edit the Relay IP address: to '10.0.0.XX' to match a physical Relay Board on my network
4. All other fields left as default
5. click the 'Add Relay' button 8 times to add relays 'new device' thru 'new device 7'
6. click the 'Add Input' button, the first click adds 'new device 8' as an input, subsequent clicks do nothing

No matter what permutation (1 relay, 8 inputs, 5 relays, 4 inputs, and so on) I can only get 9 rows in the 'Device Group (8 Max):' grid

**Comment by turgu1** [comment]
Thanks for the info, it helped me better understand your issue.

After some tests with my friend, I can tell that the notion of Device Group doesn't seem to work as expected.

My friend managed to create as many devices as required for him using the following steps, one device at a time:

1) Create a new device using the Indigo [New...] button in the main devices list and in the dialog window then select "Simple 8 Channel Relay". A second dialog will appear. 
2) In the "Configure Simple 8 Channel Relay Board" dialog, enter the IP address, Port number, Username and Password for the board.
3) Press the [Add Relay] or [Add Input] only once and press the [Save] button.
4) In the "Create New Device" Dialog, change the name of the device to something appropriate, and press the [Edit Device Settings...] button.
5) Select the appropriate Relay or Input number from the list, then click on the [Save] button in that dialog. Then close the "Create New Device" window.

I will not close this issue until I modify the code of the plugin. It may take some time as I need to go to my friend location to test the modification. It may take a week or more to complete the update. I may get rid of the notion of Device Group...

Please let me know if this workaround is helping you to create the required devices.

**Comment by Kevin-Ford** [comment]
This workaround is helpful. 

 I wonder how the developer intends the Device Group objects to work?  I assumed the Device Group would contain all the Devices encapsulated in one physical device.  (The multi sensors work this way). For consistency of structure I hope you can make these Relay boards work this way.

**Comment by Kevin-Ford** [comment]
Hi

I asked Matt at indigo (as I though this behaviour may be specific to my system) and he replied:
"
I glanced at the plugin and it is definitely coded to only allow 8 devices to be added:

    def _add_sensor(self, values, dev_id_list):
        """ Devices.xml Callback Method to add a new Relay sub-device. """
        if len(dev_id_list)-len(values["removedDevices"].split(",")) >= 8:
            return values
        ...

    def _add_relay(self, values, dev_id_list):
        """ Devices.xml Callback Method to add a new Relay sub-device. """
        if len(dev_id_list)-len(values["removedDevices"].split(",")) >= 8:
            return values
        ...

I don’t know the capabilities of the hardware, but it is fine from Indigo’s perspective to have a mix of relay and sensor devices in the same group. Normally a single group will represent a single physical module, but there are exceptions (some plugins don’t support device groups at all).

Matt”

**Comment by Kevin-Ford** [comment]
any progress on this?

**Comment by turgu1** [comment]
I’m afraid I will be able to do the debugging of the new version only by mid-september. For that, I need to borrow the Indigo environment from my friend that will not be available until then.
